kythe/rust/indexer/src/indexer/entries.rs

```diff
@@ -115,30 +116,43 @@ impl<'a> EntryEmitter<'a> {
     ///
     /// # Errors
     /// If an error occurs while writing the entry, an error is returned.
-    pub fn emit_diagnostic<S: Into<String>>(
+    pub fn emit_diagnostic(
         &mut self,
         source_vname: &VName,
-        diagnostic_vname: &VName,
         message: &str,
-        details: Option<S>,
-        url: Option<S>,
+        details: Option<&str>,
+        url: Option<&str>,
     ) -> Result<(), KytheError> {
+        // Combine diagnostic message fields and create sha256 sum
+        let combined_description =
+            format!("{}||{}||{}", message, details.unwrap_or("NONE"), url.unwrap_or("NONE"));
+        let mut sha256 = Sha256::new();
+        sha256.update(combined_description.as_str().as_bytes());
+        let bytes = sha256.finalize();
+        let sha256sum = hex::encode(bytes);
+
+        // Use source_vname signature and sha256 sum to generate diagnostic signature
+        let mut diagnostic_vname = VName::new();
+        let source_signature = source_vname.get_signature();
+        diagnostic_vname.set_signature(format!("{}_{}", source_signature, sha256sum));
+
+        // Emit diagnostic node
         self.emit_node(&diagnostic_vname, "/kythe/node/kind", b"diagnostic".to_vec())?;
```

Comment:
This had me confused for a minute - it reads like you're emitting the node multiple times. `emit_node` is really `emit_fact`.

Reply:
I agree. Not sure why I originally named it that. Changed